### DISCUSSION TITLE
Prettier for HTML options not applied

### DISCUSSION THREAD

**Discussion by mkastner** [helix-editor/helix#4924] [General] [answered]
I have these language sections in my language.toml:

```
[[language]]
name = "html"
formatter = { command = 'prettier', args = ["--parser", "html"] }
auto-format = false 

[[language]]
name = "javascript"
formatter = { command = 'prettier', args = ["--parser", "typescript"] }
auto-format = false

```
and this .prettierrc.json in my project root:

```
{
  "singleQuote": true,
  "trailingComma": "es5",
  "overrides": [
    {
      "files": ["*.hbs", "*.handlebars", "*.html"],
      "options": {
        "singleAttributePerLine": true
      }
    }
  ]
}
```

While on the one hand fmt/format on javascript  actually makes use of the .prettierrc.json settings on the other hand fmt/format on html files does not. E.g. I can't get fmt/format to apply singleAttributePerLine.

There seems to be nothing wrong with the .prettierrc.json, because it works well with neovim. It also works well for the javascript settings, e.g. toggling single and double quotes via singleQuote property in .prettierrc.json.

Invoking prettier on command line on this html:

```
<!-- file test.html -->
<div id="1" class="test">asdf</div>
```

`> prettier test.html`

properly generates

```
<!-- file test.html -->
<div
  id="1"
  class="test"
>
  test
</div>
```

But helix doesn't apply singleAttributePerLine.

Can anyone help me with this?

**Comment by David-Else** [comment]
Maybe you are not reading the config file for HTML and it is using the LSP formatter? Providing your current working directory is the root of the project (and Linux/Mac `./` syntax), how about:
```
formatter = { command = 'prettier', args = ["--parser", "typescript", "--config", "./prettierrc.json"] }
```
Based on https://prettier.io/docs/en/cli.html#--find-config-path-and---config , I have not tried it.

**Comment by mkastner** [comment]
@David-Else Thanks for your reply. I've tried what you proposed it and it's still not applying the settings of .prettierrc.json. BTW the relative path would be "./.prettierrc.json" since it's a dot file.
And as I've written, the settings for javascript are being picked up .prettierrc.json with similar settings. I haven't specified the path to .prettierrc.json there either.

**Comment by David-Else** [comment]
It seems very mysterious, my suspicion is still with the config file not being loaded, how about using the absolute path just to be certain, and if that doesn't work then a broken path to see how it reacts? I have seen Helix need an absolute path on Windows as a bug in the past.

**Comment by mkastner** [comment]
@David-Else I've tried the absolute path: didn't work either. I'm on a Mac.

Running

`> hx --health html`

produces

```
Configured language server: vscode-html-language-server
Binary for language server: /usr/local/bin/vscode-html-language-server
Configured debug adapter: None
Highlight queries: ✓
```

I'm not sure, if prettier should appear there. In any case, it doesn't.



I've checked

`~/.cache/helix/helix.log`

and this is what I see there:

```
2022-11-28T18:03:46.496 helix_term::application [ERROR] Language Server: Method not found client/registerCapability
2022-11-28T18:07:09.666 helix_view::editor [ERROR] Failed to initialize the LSP for `source.toml` { cannot find binary path }
```

It looks like there's something not properly working with the language server access.

**Reply by gabydd** [reply]
Hmm I wonder if you have prettier globally installed or if it's local to the project maybe prefix the command with ``npx ``

**Reply by gabydd** [reply]
Sorry the command should be npx and prettier should be an arg to it

**Reply by gabydd** [reply]
Probably can disregard my comment, my specific thing is a solution to a problem I was having actually I think

**Comment by David-Else** [comment, marked as answer]
I don't think Prettier should appear there at this time, but it could be a good feature request.

It looks like the `["--parser", "html"] ` tells Helix it is HTML, but `"files": [ "*.html"],` is not able to recognise the buffer's file extension for `"overrides"`. 

I suggest you look at the original commit that added the external format using a binary: https://github.com/helix-editor/helix/pull/2942 , and consider asking the author of the PR for help.

**Comment by mkastner** [comment]
@David-Else 

thanks again for your reply.

`ls -lah /usr/local/bin/prettier`

produces

`/Users/[username]/yarn/global/node_modules/.bin/prettier`

I don't think there's anything wrong there.

Invoking (in my project's root)

prettier test.html

works and formats the tag attributes properly. The conclusion I draw from this, is that "files": [ "*.html" ] is properly interpreted.

I'll try to check again, I have screwed something up. Maybe then I'll contact https://github.com/helix-editor/helix/pull/2942 's author.

**Reply by trey** [reply]
Did you ever find a solution to this? I'm having the same issue. Helix picks up my project's `.prettierrc`, but it won't parse the `overrides` section.

```json
{
  "useTabs": true,
  "tabWidth": 2,
  "singleQuote": true,
  "overrides": [
    {
      "files": "*.css",
      "options": {
        "singleQuote": false
      }
    }
  ]
}
```

**Reply by satoqz** [reply]
I was able to get this working by passing `--stdin-filepath` to prettier. For example, for CSS:

```toml
[[language]]
name = "css"
formatter = { command = "prettier", args = ["--parser", "css", "--stdin-filepath", "dummy.css"] }
auto-format = true
```

where dummy.css doesn't need to exist, and can be anything that matches the override.
